lib/ansible/modules/cloud/opennebula/one_service.py

```diff
@@ -0,0 +1,762 @@
+
+    if service:
+        return service["ID"]
+
+    return None
+
+
+def get_connection_info(module):
+
+    url = module.params.get('api_url')
+    username = module.params.get('api_username')
+    password = module.params.get('api_password')
+
+    if not url:
+        url = os.environ.get('ONEFLOW_URL')
+
+    if not username:
+        username = os.environ.get('ONEFLOW_USERNAME')
+
+    if not password:
+        password = os.environ.get('ONEFLOW_PASSWORD')
+
+    if not(url and username and password):
+        module.fail_json(msg="One or more connection parameters (api_url, api_username, api_password) were not specified")
+    from collections import namedtuple
+
+    auth_params = namedtuple('auth', ('url', 'user', 'password'))
+
+    return auth_params(url=url, user=username, password=password)
+
+
+def main():
+    fields = {
+        "api_url": {"required": False, "type": "str"},
+        "api_username": {"required": False, "type": "str"},
+        "api_password": {"required": False, "type": "str", "no_log": True},
+        "service_name": {"required": False, "type": "str"},
+        "service_id": {"required": False, "type": "int"},
+        "template_name": {"required": False, "type": "str"},
+        "template_id": {"required": False, "type": "int"},
+        "state": {
+            "default": "present",
+            "choices": ['present', 'absent'],
+            "type": "str"
+        },
+        "mode": {"required": False, "type": "str"},
+        "owner_id": {"required": False, "type": "int"},
+        "group_id": {"required": False, "type": "int"},
+        "unique": {"default": False, "type": "bool"},
+        "wait": {"default": False, "type": "bool"},
+        "wait_timeout": {"default": 300, "type": "int"},
+        "custom_attrs": {"default": {}, "type": "dict"},
+        "role": {"required": False, "type": "str"},
+        "cardinality": {"required": False, "type": "int"},
+        "force": {"default": False, "type": "bool"}
+    }
+
+    module = AnsibleModule(argument_spec=fields, supports_check_mode=True)
+
+    auth = get_connection_info(module)
+    params = module.params
+    service_name = params.get('service_name')
+    service_id = params.get('service_id')
+    template_name = params.get('template_name')
+
+    template_id = params.get('template_id')
+    state = params.get('state')
+    permissions = params.get('mode')
+    owner_id = params.get('owner_id')
+    group_id = params.get('group_id')
+    unique = params.get('unique')
+    wait = params.get('wait')
+    wait_timeout = params.get('wait_timeout')
+    custom_attrs = params.get('custom_attrs')
+    role = params.get('role')
+    cardinality = params.get('cardinality')
+    force = params.get('force')
+
+    if template_id:
+        if not get_template_by_id(module, auth, template_id):
+            module.fail_json(msg="There is no template with template_id: " + str(template_id))
+
+    if template_id and template_name:
```

Comment:
not a blocker but there are many helper codes for mutually_exclusive, required_if, required_one_of etc. args in AnsibleModule